pkg/sqlite/blob.go

```diff
@@ -0,0 +1,384 @@
+
+	if qb.options.UseFilesystem {
+		ret, err := qb.fsStore.Read(ctx, checksum)
+		if err != nil {
+			if !errors.Is(err, fs.ErrNotExist) {
+				return nil, checksum, fmt.Errorf("reading from filesystem: %w", err)
+			}
+
+			// not found in the filesystem - should not happen
+			return nil, checksum, &ChecksumFileNotExistError{
+				Checksum: checksum,
+			}
+		}
+
+		return ret, checksum, nil
+	}
+
+	return nil, checksum, fmt.Errorf("unexpected nil blob")
+}
+
+// Read reads the data from the database or filesystem, depending on which is enabled.
+func (qb *BlobStore) Read(ctx context.Context, checksum string) ([]byte, error) {
+	if !qb.options.UseDatabase && !qb.options.UseFilesystem {
+		panic("no blob store configured")
+	}
+
+	// check the database first
+	ret, err := qb.read(ctx, checksum)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return nil, fmt.Errorf("reading from database: %w", err)
+		}
+
+		// not found in the database - does not exist
+		return nil, &ChecksumNotFoundError{
+			Checksum: checksum,
+		}
+	}
+
+	if qb.options.UseDatabase && ret != nil {
+		return ret, nil
+	}
+
+	if qb.options.UseFilesystem {
+		ret, err := qb.fsStore.Read(ctx, checksum)
+		if err != nil {
+			if !errors.Is(err, fs.ErrNotExist) {
+				return nil, fmt.Errorf("reading from filesystem: %w", err)
+			}
+
+			// not found in the filesystem - should not happen
+			return nil, &ChecksumFileNotExistError{
+				Checksum: checksum,
+			}
+		}
+
+		return ret, nil
+	}
+
+	return nil, fmt.Errorf("unexpected nil blob")
+}
+
+func (qb *BlobStore) read(ctx context.Context, checksum string) ([]byte, error) {
+	q := dialect.From(qb.table()).Select(qb.table().All()).Where(qb.tableMgr.byID(checksum))
+
+	var row blobRow
+	const single = true
+	if err := queryFunc(ctx, q, single, func(r *sqlx.Rows) error {
+		if err := r.StructScan(&row); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("querying %s: %w", qb.table(), err)
+	}
+
+	return row.Blob, nil
+}
+
+// Delete marks a checksum as no longer in use by a single reference.
+// If no references remain, the blob is deleted from the database and filesystem.
+func (qb *BlobStore) Delete(ctx context.Context, checksum string) error {
+	// try to delete the blob from the database
+	if err := qb.delete(ctx, checksum); err != nil {
+		if qb.isConstraintError(err) {
+			// blob is still referenced - do not delete
+			return nil
+		}
+
+		// unexpected error
+		return fmt.Errorf("deleting from database: %w", err)
+	}
+
+	// blob was deleted from the database - delete from filesystem if enabled
+	if qb.options.UseFilesystem {
+		if err := qb.fsStore.Delete(ctx, checksum); err != nil {
+			return fmt.Errorf("deleting from filesystem: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (qb *BlobStore) isConstraintError(err error) bool {
+	var sqliteError sqlite3.Error
+	if errors.As(err, &sqliteError) {
+		return sqliteError.Code == sqlite3.ErrConstraint
```

Comment:
Should be `err.Code == sqlite3.ErrConstraint` instead ? Can not actually test atm

Reply:
`err` is an `error` interface and does not have the `Code` field. This checks if `err` is an instance of `sqlite3.Error` and if so, then checks if it's `Code` field is `sqlite3.ErrConstraint`.